0005-R-pubsub-delayed-scheduling.md

```diff
@@ -0,0 +1,156 @@
+
+#### Configuring delayed messaging
+
+Users can enable delayed messaging by adding the following pub/sub component metadata (other fields specific for Redis removed for brevity):
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: pubsub
+spec:
+  type: pubsub.redis
+  version: v1
+  metadata:
+  - name: "enableDelayedMessaging"
+    value: true
+```
+
+If not set, the default value is `false` and the feature is bypassed completely.
+
+
+#### Consuming delayed messages
+
+On startup, the Dapr runtime will subscribe to a topic in the following format: `<app-id>-delayed`. This topic holds all delayed messages for the given application, and has the following benefits:
+
+1. Topics messages are scoped per application and not per component
+2. Enables infrastructure transparency in that operators can examine the underlying broker, discover the delayed messages topic and optionally look into the message queue if supported
+3. Of secondary importance, this enables users to optionally consume messages directly from the configured topic for advanced scenarios
+
+##### Interaction with namespaced consumer groups
+
+If [namespaced consumer groups](https://docs.dapr.io/developing-applications/building-blocks/pubsub/howto-namespace/#with-namespace-consumer-groups) are enabled, the topic format will result in `<app-id><namespace>-delayed` to allow for a delayed message topic that is scoped both to the app id and the namespace it is running in.
+
+#### Publishing delayed messages
+
+When Dapr receives a message that has a metadata key of `scheduledSendTime`, it will publish it to the delayed message topic in the format described above, if `enableDelayedMessaging` has been enabled on the component.
+Since this feature supports all pub/sub components, Dapr will return an error if a metadata key is present and `enableDelayedMessaging` is false or not set on the component.
+
+### Message format
+
+Dapr instances that receive a delayed message need to know the topic and other request level details (content type, etc.). To faciliate this requirement and support both CloudEvents and raw payloads, Dapr will publish a well known, typed proto message to the delayed messages topic in the following structure:
+
+```proto
+message DelayedMessage {
+  string topic = 1;
+  bytes message = 2;
+  string contentType = 3;
+  map<string, string> metadata = 4;
+  bool bulkPublish = 5;
+}
+```
+
+This provides the Dapr runtime all the information it needs to create a publish request and send it to the message broker.
+
+#### Bulk publish support
+
+If the `DelayedMessage` has the `bulkPublish` field set to `true`, the `message` field will be deserialized as a [BulkMessageEntry](https://github.com/dapr/components-contrib/blob/677663e81ab292d140021e671841ca5f613aaaea/pubsub/requests.go#L26) type and will be sent via the bulk publish route.
```

Comment:
Why is this being treated differently? 
Instead can this following flow be done? 
When a bulkPublish request is made and scheduledSendTime is set on any message in that request, that message with metadata alone is sent to the `<app>-delayed` topic, why do we need to send the whole of the BulkMessageEntry to the topic? 
On subscribing to the `<app>-delayed` topic, only the message with metadata will be there in terms of cloudevent or for raw payload only the message payload, that alone can then be forwarded to the correct topic.

Main  question is in the `<app>-delayed` topic, why should we serialize and store BulkMessageEntry? Why can't we just store the Dapr created CE/ User input CE/ raw payload directly even for bulkPublish?

Reply:
We could, but looking at the code currently in master branch it looks like the publish bulk method in contrib expects a BulkMessageEntry. 
This is more of an implementation detail.